pkg/webhook/admission/pod/metrics_aggregate_injector.go

```
@@ -17,46 +17,91 @@ limitations under the License.
 package pod
 
 import (
+	"encoding/json"
 	"github.com/kserve/kserve/pkg/constants"
 	v1 "k8s.io/api/core/v1"
 )
 
-const defaultPrometheusPort = "8080"
+const (
+	defaultKserveContainerPrometheusPort = "8080"
+	MetricsAggregatorConfigMapKeyName    = "metricsAggregator"
+)
 
-// InjectMetricsAggregator looks for the annotations to enable aggregate kserve-container and queue-proxy metrics and
-// if specified, sets port-related EnvVars in queue-proxy and the aggregate prometheus annotation.
-func InjectMetricsAggregator(pod *v1.Pod) error {
+type MetricsAggregator struct {
+	//TODO: check configmap.go, do we need omitfalse & the +kubebuilder:object:generate=false above struct?
+	EnableMetricAggregation           string `json:"enableMetricAggregation"`
+	EnablePrometheusAggregateScraping string `json:"enablePrometheusAggregateScraping"`
+}
+
+func newMetricsAggregator(configMap *v1.ConfigMap) (*MetricsAggregator, error) {
+	ma := &MetricsAggregator{}
+
+	if maConfigVal, ok := configMap.Data[MetricsAggregatorConfigMapKeyName]; ok {
+		err := json.Unmarshal([]byte(maConfigVal), &ma)
+		//TODO: should I also panic here? seems weird that the other injectors do.
```

I notice we panic in the other injectors after this unmarshal line of code but I don't get why because we can return err also. we could create a common error type and use that - so I guess my q is - is there a reason we want to panic after unmarshal instead of return err ? if so, should we remove the return err from each of these methods?

---

if we do not panic on bad configuration and continue we may leave the controller in a bad state depending on some unexpected configurations. We'd rather let the program crash and not continue.

---

ok that is fine, we probably shouldn't return an err then?

---

I actually forgot to add a validation check here that can return err. so I will need it in the func